Add Nvidia GPU TEE Support

jax-cn opened the pull request.

Alex-yang00 committed:
fix: update SNP release URL to version 5.19.0

Alex-yang00 committed:
feat: add GPU setup functionality and download GPU admin tools

- Introduced a new setup_gpu function to configure GPU CC for the host system.
- Added command to download GPU admin tools from the NVIDIA repository during initialization.
- Updated main function to include the new setup_gpu command.

Alex-yang00 committed:
feat: enhance GPU setup with passthrough configuration

- Updated setup_gpu function to include GPU passthrough configuration.
- Added a new script, gpu_passthrough.sh, to manage NVIDIA GPU passthrough for virtual machines.
- Implemented functions for detecting GPUs, enabling VFIO-PCI, and configuring passthrough.

Alex-yang00 committed:
feat: add NVIDIA modprobe configuration for GPU support

- Created nvidia-lkca.conf to manage NVIDIA module loading with ecdsa_generic and ecdh.
- Ensured proper permissions for the configuration file.

Alex-yang00 committed:
feat: enhance launch script and configuration for GPU support

- Updated VCPU count limit from 255 to 32 in launch.sh.
- Modified UEFI drive options for improved compatibility.
- Added GPU passthrough parameters to support NVIDIA GPUs in launch.sh.
- Adjusted memory configuration in config.py from 204800 to 32768.
- Enhanced base_setup.sh to install NVIDIA drivers and CUDA toolkit.
- Increased tmpfs sizes for /var and /tmp in init.sh for better resource management.

Alex-yang00 committed:
fix: update hb_branch to use 'main' and change SNP release URL to version 6.9.0

Alex-yang00 committed:
feat: add flags to enable/disable gpu

Alex-yang00 committed:
fix: configure base_setup by gpu flag

PeterFarber commented on launch.sh:
```diff
@@ -386,7 +393,7 @@ else
 
 # add number of VCPUs
-[ -n "${SMP}" ] && add_opts "-smp ${SMP},maxcpus=255"
+[ -n "${SMP}" ] && add_opts "-smp ${SMP},maxcpus=32"
```
Why is this nessasary?

PeterFarber commented on run:
```diff
@@ -61,13 +61,15 @@ def init():
     # Download and extract SNP release tarball.
     tarball = os.path.join(config.dir.build, "snp-release.tar.gz")
-    url = "https://github.com/SNPGuard/snp-guard/releases/download/v0.1.2/snp-release.tar.gz"
+    url = "https://github.com/apuslabs/snp-guard/releases/download/v6.9.0/snp-release.tar.gz"
```
Will upload release to ardrive and fork this repo to permaweb. 
Then use the arweave tx to download

PeterFarber commented on run:
```diff
@@ -61,13 +61,15 @@ def init():
     run_command(f"tar -xf {tarball} -C {config.dir.build}")
     run_command(f"rm {tarball}")
+    # Download the GPU admin tools
+    run_command(f"cd {config.dir.build} && git clone https://github.com/nvidia/gpu-admin-tools && cd .. ")
```
I will work the gpu-admin-tools to permaweb.
Then replace this with the forked repo.

PeterFarber commented on scripts/base_setup.sh:
```diff
@@ -9,6 +12,34 @@ node -v && npm -v
+    export NEEDRESTART_MODE=a
+
+    echo "Installing NVIDIA CUDA repository..."
+    wget https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64/cuda-keyring_1.1-1_all.deb
```
Download this and upload to ardrive and use an arweave tx id to download.

PeterFarber committed:
Switch to edge branch and migrate dependencies to Arweave

- Update HyperBEAM branch from main to edge
- Use dynamic SMP value for QEMU maxcpus instead of hardcoded 32
- Replace GitHub/NVIDIA URLs with Arweave for SNP release and CUDA keyring
- Switch gpu-admin-tools repository from nvidia to permaweb
- Add GPU setup integration to base image build process
